File: generated_code/python/solution_2903_1.py
from typing import Optional
import math

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # If the list is empty or has only one node, no adjacent pairs exist.
        if not head or not head.next:
            return head
        
        # Use a pointer to traverse the list and inspect adjacent pairs.
        curr = head
        while curr and curr.next:
            # Calculate the GCD of the current node and the next node.
            gcd_val = math.gcd(curr.val, curr.next.val)
            
            # Create the new node to insert.
            new_node = ListNode(gcd_val)
            
            # Insert new_node between curr and curr.next.
            new_node.next = curr.next
            curr.next = new_node
            
            # Move the pointer past the newly inserted node to the original next node.
            curr = new_node.next
            
        return head

File: generated_code/cpp/solution_2903_1.cpp
#include <numeric> // For std::gcd

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    /**
     * Algorithm: Linear Traversal with In-place Insertion
     * 
     * We traverse the linked list using a pointer 'curr'. For each node, 
     * if a 'next' node exists, we calculate the GCD of the current and 
     * next node values. We then create a new node with this GCD and 
     * insert it between them.
     * 
     * Time Complexity: O(N * log(min(V))), where N is the number of nodes 
     * and V is the maximum value in a node (log factor due to GCD calculation).
     * Space Complexity: O(1) auxiliary space (ignoring the space for new nodes).
     */
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        // If the list is empty or has only one node, no pairs exist to insert GCDs.
        if (!head || !head->next) {
            return head;
        }

        ListNode* curr = head;

        // Traverse until the second to last node
        while (curr != nullptr && curr->next != nullptr) {
            // Calculate GCD of the two adjacent nodes using C++17 std::gcd
            int commonDivisor = std::gcd(curr->val, curr->next->val);

            // Create the new node and link it: curr -> newNode -> nextNode
            ListNode* newNode = new ListNode(commonDivisor, curr->next);
            curr->next = newNode;

            // Move curr forward by two steps: Skip the newly inserted node 
            // to reach the original "next" node for the next iteration.
            curr = newNode->next;
        }

        return head;
    }
};